src/libcore/str/pattern.rs

```diff
@@ -596,6 +607,17 @@ impl<'a, 'b> StrSearcher<'a, 'b> {
 unsafe impl<'a, 'b> Searcher<'a> for StrSearcher<'a, 'b> {
     fn haystack(&self) -> &'a str { self.haystack }
 
+    fn as_str(&self) -> &'a str {
+        match self.searcher {
+            StrSearcherImpl::Empty(ref searcher) => {
+                &self.haystack[searcher.position..searcher.end]
+            },
+            StrSearcherImpl::TwoWay(ref searcher) => {
+                &self.haystack[searcher.position..searcher.end]
```

Comment:
For example, position and end have no relation and position > end is a legal state.